feat(scan): provide a default scan factory

ostridm opened the pull request:
closes #68

derevnjuk commented on src/SecTester.Scan/Target/Har/Response.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace SecTester.Scan.Target.Har;
+
+public record Response(int Status, string StatusText, string HttpVersion, IEnumerable<Cookie> Cookies,
```
`Response` should extend `Message`. It is pointless to add all the fields into constructors, we have already discussed this. The fields like `Cookies` or `Headers` should have a default value

derevnjuk commented on src/SecTester.Scan/Target/Har/Entry.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace SecTester.Scan.Target.Har;
+
+public record Entry(DateTime StartedDateTime, int Time, Timings Timings, Cache Cache, Request Request,
```
ditto

derevnjuk commented on src/SecTester.Scan/Target/Har/Log.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace SecTester.Scan.Target.Har;
+
+public record Log(Tool Creator, IEnumerable<Entry> Entries, string Version = "1.2")
```
ditto

derevnjuk commented on src/SecTester.Scan/Target/Har/Timings.cs:
```diff
@@ -0,0 +1,8 @@
+namespace SecTester.Scan.Target.Har;
+
+public record Timings(int Wait, int Receive)
```
ditto

derevnjuk commented on src/SecTester.Scan/Target/Har/Timings.cs:
```diff
@@ -0,0 +1,8 @@
+
+public record Timings(int Wait, int Receive)
+{
+  public int? Send { get; init; }
```
The `Send` is required like `Wait` or `Receive`, but all of them have default values

ostridm replied:
`Send` is optional in `har-sdk`
https://github.com/NeuraLegion/har-sdk/blob/10baaf123200ad80d0262778c4069fe82d28727d/packages/core/src/types/har.ts#L818

derevnjuk replied:
You can refer to whatever you want, but this has nothing in common with the spec http://www.softwareishard.com/blog/har-12-spec/

derevnjuk commented on src/SecTester.Scan/Target/Har/Timings.cs:
```diff
@@ -0,0 +1,8 @@
+{
+  public int? Send { get; init; }
+
+  public static readonly Timings Default = new(0, 0) { Send = 0 };
```
It is pointless to have `Default` static field, see comment above

derevnjuk commented on test/SecTester.Scan.Tests/Extensions/HttpContentExtensions.cs:
```diff
@@ -0,0 +1,11 @@
+namespace SecTester.Scan.Tests.Extensions;
+
+internal static class HttpContentExtensions
```
I didn't get it, why do all your PR contain the same changes? How you are going to resolve conflicts? Why do these changes relate to this PR?

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+    string? name,
+    bool? smart, int? poolSize, TimeSpan? targetTimeout,
+    TimeSpan? slowEpTimeout, bool? skipStaticParams,
+    IEnumerable<AttackParamLocation>? attackParamLocations)
```
According to the interface, almost all parameters are optional, why do you need to introduce this huge constructor? The constructor should contain only the required parameters, everything else should be initialized using object initializers

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+
+  private static readonly IEnumerable<AttackParamLocation> AttackParamLocationWhiteList = Enum
+    .GetValues(typeof(AttackParamLocation))
+    .Cast<AttackParamLocation>();
```
You should use `Enum.IsDefined(typeof(AttackParamLocation), value)`

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+    return name.Length <= MaxNameLength ? name : name.Substring(0, MaxNameLength);
+  }
+
+  private static void AssertEnum<T>(IEnumerable<T>? value, IEnumerable<T> whiteList,
```
Such methods should never be an instance or class member. Also, I don't see any reason to have so a weak method at all: unclear signature, irrelevant name, etc

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+    .Cast<AttackParamLocation>();
+
+  private Target.Target _target = null!;
+  private IEnumerable<TestType> _tests = null!;
```
Why do you initialize them with null?

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+    set
+    {
+      if (string.IsNullOrWhiteSpace(value) ||
+          value!.Length is < MinNameLength or > MaxNameLength)
```
`string.IsNullOrWhiteSpace(value)` and `value!.Length is < MinNameLength` do the same job

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+  private static string CreateDefaultName(TargetOptions target)
+  {
+    var uri = new Uri(target.Url);
+    var name = $"{target.Method ?? HttpMethod.Get} {uri.Host}";
```
ditto

derevnjuk replied:
ditto

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+    AttackParamLocations = attackParamLocations ?? DefaultAttackParamLocations;
+  }
+
+  private static string CreateDefaultName(TargetOptions target)
```
`TargetOptions` shouldn't be used to retrieve `Url` (URL must be normalized). It is already done in `Target`

derevnjuk replied:
`TargetOptions` shouldn't be used to retrieve `Url` (URL must be normalized). It is already done in `Target`

derevnjuk replied:
`TargetOptions` shouldn't be used to retrieve `Url` (URL must be normalized). It is already done in `Target`

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+    var uri = new Uri(target.Url);
+    var name = $"{target.Method ?? HttpMethod.Get} {uri.Host}";
+
+    return name.Length <= MaxNameLength ? name : name.Substring(0, MaxNameLength);
```
The last symbol must be replace with `…` once name is truncated

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+  public string? Name
+  {
+    get { return _name; }
+    set
```
Each accessible field must be init only https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023926331

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+  private const int MaxNameLength = 200;
+  private const int MinPoolSize = 1;
+  private const int MaxPoolSize = 50;
+  private const bool DefaultSmart = true;
+  private const bool DefaultSkipStaticParams = true;
+  private const int DefaultPoolSize = 10;
```
Why do you introduce these static fields? Why don't assign default values directly to the accessible fields?

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+  private static readonly TimeSpan MinTargetTimeout = TimeSpan.FromSeconds(1);
+  private static readonly TimeSpan MaxTargetTimeout = TimeSpan.FromSeconds(120);
+  private static readonly TimeSpan MinSlowEpTimeout = TimeSpan.FromSeconds(100);
+  private static readonly TimeSpan DefaultTargetTimeout = TimeSpan.FromSeconds(5);
+  private static readonly TimeSpan DefaultSlowEpTimeout = TimeSpan.FromSeconds(1000);
+  private static readonly IEnumerable<AttackParamLocation> DefaultAttackParamLocations = new List<AttackParamLocation>
+  {
+    AttackParamLocation.Body, AttackParamLocation.Query, AttackParamLocation.Fragment
+  };
```
ditto

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,199 @@
+
+  public string? Name
+  {
+    get { return _name; }
```
Use expression body where it is possible
```suggestion
    get => _name;
```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -22,14 +23,18 @@ public class Scan : IDisposable
+    Id = id ?? throw new ArgumentNullException(nameof(id));
     _scans = scans ?? throw new ArgumentNullException(nameof(scans));
     _options = options ?? throw new ArgumentNullException(nameof(options));
+    _logger = logger ?? throw new ArgumentNullException(nameof(logger));
```
You don't use the logger, why did you add it here?

derevnjuk commented on src/SecTester.Scan/Target/Har/Response.cs:
```diff
@@ -0,0 +1,18 @@
+  public IEnumerable<Header> Headers { get; } = Headers ?? throw new ArgumentNullException(nameof(Headers));
+  public Content Content { get; } = Content ?? throw new ArgumentNullException(nameof(Content));
+
+  public static readonly Response Default = new(200, "OK", "HTTP/1.1", new List<Cookie>(),
+    new List<Header>(), "", -1, -1, Content.Default);
```
This is the default value for `ScanFactory`, but it has nothing in common with the default value for HAR spec in general

derevnjuk commented on src/SecTester.Scan/DefaultScanFactory.cs:
```diff
@@ -0,0 +1,98 @@
+
+public class DefaultScanFactory : ScanFactory
+{
+  private const int HarDefaultNameMaxHostLength = 200;
```
Har default name max host length? What? 
```suggestion
  private const int MaxSlugLength = 200;
```

derevnjuk commented on src/SecTester.Scan/DefaultScanFactory.cs:
```diff
@@ -0,0 +1,98 @@
+    return new Scan(scanId, _scans, _logger, options ?? new ScanOptions());
+  }
+
+  private async Task<ScanConfig> BuildScanConfig(ScanSettingsOptions settingsOptions)
```
```suggestion
  private async Task<ScanConfig> BuildScanConfig(ScanSettings settings)
```

derevnjuk commented on src/SecTester.Scan/DefaultScanFactory.cs:
```diff
@@ -0,0 +1,98 @@
+
+    host = host.Length <= HarDefaultNameMaxHostLength ? host : host.Substring(0, HarDefaultNameMaxHostLength);
+
+    return $"{host.TrimEnd('-')}-{Guid.NewGuid().ToString()}.har";
```
```suggestion
    return $"{host.TrimEnd('-')}-{Guid.NewGuid()}.har";
```

derevnjuk reviewed (changes requested):
BTW comment in Issue and comment in Har are not the same thing

derevnjuk commented on test/SecTester.Scan.Tests/Mocks/MockLogger.cs:
```diff
@@ -0,0 +1,19 @@
+namespace SecTester.Scan.Tests.Mocks;
+
+internal abstract class MockLogger : ILogger
```
There are a lot of duplicates from your previous PR. How do you want me to review them? https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1038232222

ostridm replied:
@derevnjuk Nope, I am going to rebase this PR after #59 is merged. Convert it to draft for a while.

derevnjuk reviewed (changes requested):
Please describe how to use the project in README

ostridm committed:
feat(scan): provide a default scan factory

update README.md, adjust according to code review

closes #68

derevnjuk commented on test/SecTester.Scan.Tests/Commands/UploadHarTests.cs:
```diff
@@ -4,7 +4,10 @@ public class UploadHarTests
-  private static readonly Har Har = new();
+  private static readonly Har Har = new(
+    new Log(
+      new Tool("Configuration_Name", "Configuration_Version"))
```
Avoid using any cryptic and weird test data
```suggestion
      new Tool("name", "v1.1.1"))
```

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -17,6 +17,20 @@ public class StringUtilsTests
+    new object[] { "aa", "…", 0 },
+    new object[] { "aa", "a…", 1 },
+    new object[] { "aa", "aa", 2 },
+    new object[] { "aa", "aa", 3 }
```
What is difference between this `"a", "a", 1` and these sets? Remove similar test cases. Instead of inventing bicycle, just reuse described in sectester-js

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -27,4 +41,15 @@ public void ToSnakeCase_SatisfyConversion(string input, string expected)
+
+  [Theory]
+  [MemberData(nameof(TruncateMapping))]
+  public void Truncate_SatisfyConversion(string input, string expected, int length)
```
`Length` is also input) The argument name are confusing

derevnjuk commented on src/SecTester.Core/Utils/StringUtils.cs:
```diff
@@ -21,6 +20,12 @@ public static class StringUtils
         .ToLowerInvariant();
+  }
 
+  public static string? Truncate(this string? value, int n)
```
It is not clear why you accept nullable input, this is weak design

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -17,6 +17,20 @@ public class StringUtilsTests
     new object[] { "kebab-case_with_snake", "kebab-case_with_snake" },
   };
 
+  public static readonly IEnumerable<object[]> TruncateMapping = new List<object[]>
```
Mapping?

derevnjuk commented on src/SecTester.Core/Utils/StringUtils.cs:
```diff
@@ -21,6 +20,12 @@ public static class StringUtils
+  {
+    return string.IsNullOrEmpty(value) || n < 0 || value!.Length <= n
+      ? value
+      : $"{value.Substring(0, n)}…";
```
Invert condition to avoid redundant checks and uncertainty

derevnjuk replied:
Your implementation work is unexpected, you violate the postcondition.

derevnjuk commented on src/SecTester.Scan/README.md:
````diff
@@ -14,6 +14,250 @@ abstraction to handle events and commands.
 $ dotnet add package SecTester.Scan
 ```
 
+## Usage
````
There are ton of formatting issues and oddities in both code snippets and text.

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  [Fact]
+  public void PublicConstructor_CreatesInstanceWithDefaultOptions()
```
```suggestion
  public void ScanSettings_CreatesInstanceWithDefaultOptions()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  [Fact]
+  public void InternalConstructor_CreatesInstance()
```
```suggestion
  public void ScanSettings_ScanSettingsOptionsPassed_CreatesInstance()
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  public void Dispose()
+  {
+    _targetOptions.ClearSubstitute();
+    _scanSettingsOptions.ClearSubstitute();
```
```suggestion
    _scanSettingsOptions.ClearSubstitute();
    GC.SuppressFinalize(this);
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+    _scanSettingsOptions.SlowEpTimeout.Returns(TimeSpan.FromSeconds(200));
+    _scanSettingsOptions.SkipStaticParams.Returns(false);
+    _scanSettingsOptions.Tests.Returns(new List<TestType> { TestType.Csrf });
+    _scanSettingsOptions.AttackParamLocations.Returns(new List<AttackParamLocation> { AttackParamLocation.Query });
```
It is bad practice to use mock or stubs for test data. Moreover, this is pointless

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+
+  public static IEnumerable<object[]> SetterInvalidInput()
+  {
+    yield return new object[]
```
Why do you use custom iterator instead of a list like in other places??

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+
+  [Theory]
+  [MemberData(nameof(SetterInvalidInput))]
+  public void Setter_GivenInvalidInput_ThrowError(string expectedMessage,
```
```suggestion
  public void Setter_GivenInvalidInput_ThrowsError(string expected,
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  [Theory]
+  [MemberData(nameof(SetterInvalidInput))]
+  public void Setter_GivenInvalidInput_ThrowError(string expectedMessage,
+    Func<ScanSettings, ScanSettings> action)
```
This is a mutator or test data factory, but exactly not action.

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  [Fact]
+  public void InternalConstructor_CreatesInstanceWithUniqueTests()
```
ditto (you want to say that this case is not applicable for other constructors?)

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  [Fact]
+  public void InternalConstructor_CreatesInstanceWithUniqueAttackParamLocations()
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+    var result = new ScanSettings(_scanSettingsOptions);
+
+    // assert
+    result.Should().BeEquivalentTo(new { Name = $"GET {new string('a', 195)}…" });
```
Should we tune this value each time when something is changed?

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  [Fact]
+  public void InternalConstructor_NameIsNull_CreatesInstanceTruncatedHost()
```
What about the case with non-default name that is greater than 200 characters?

derevnjuk commented on test/SecTester.Scan.Tests/ScanSettingsTests.cs:
```diff
@@ -0,0 +1,237 @@
+  public void InternalConstructor_NameIsNull_CreatesInstanceTruncatedHost()
+  {
+    // arrange
+    _scanSettingsOptions.Name.ReturnsForAnyArgs(_ => null);
```
Why do you use `ReturnsForAnyArgs` if this is a property that definitely does not accept any arguments?

derevnjuk commented on test/SecTester.Scan.Tests/Models/UploadHarOptionsTests.cs:
```diff
@@ -1,30 +1,30 @@
+
+  private static readonly Har Har = new(
+    new Log(
+      new Tool("Configuration_Name", "Configuration_Version")
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/Models/UploadHarOptionsTests.cs:
```diff
@@ -1,30 +1,30 @@
-  private readonly Har _har = new();
+  private const string HarFileName = "file.har";
+
+  private static readonly Har Har = new(
```
Avoid using static test data

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -46,7 +45,11 @@ public class DefaultScansTests : IDisposable
-  private static readonly Har Har = new();
+  private static readonly Har Har = new(
+    new Log(
+      new Tool("Configuration_Name", "Configuration_Version")
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScansTests.cs:
```diff
@@ -46,7 +45,11 @@ public class DefaultScansTests : IDisposable
   { Cvss = "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:L" };
 
-  private static readonly Har Har = new();
+  private static readonly Har Har = new(
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScanFactoryTests.cs:
```diff
@@ -0,0 +1,100 @@
+  private const string ScanId = "roMq1UVuhPKkndLERNKnA8";
+
+  private readonly Configuration _configuration = new("app.neuralegion.com");
+  private readonly ScanSettingsOptions _options = Substitute.For<ScanSettingsOptions>();
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScanFactoryTests.cs:
```diff
@@ -0,0 +1,100 @@
+  public async Task CreateScan_GeneratesUploadHarFile()
+  {
+    // arrange
+    _options.Name.ReturnsForAnyArgs(null as string);
+    _options.AttackParamLocations.ReturnsForAnyArgs(null as IEnumerable<AttackParamLocation>);
```
Why is it necessary? That has nothing in common with this particular test case

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScanFactoryTests.cs:
```diff
@@ -0,0 +1,100 @@
+  public async Task CreateScan_TruncatesHarFilename()
+  {
+    // arrange
+    _options.Name.ReturnsForAnyArgs(null as string);
+    _options.AttackParamLocations.ReturnsForAnyArgs(null as IEnumerable<AttackParamLocation>);
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/DefaultScanFactoryTests.cs:
```diff
@@ -0,0 +1,100 @@
+  public async Task CreateScan_CreatesScan()
+  {
+    // arrange
+    _options.Name.ReturnsForAnyArgs(null as string);
+    _options.AttackParamLocations.ReturnsForAnyArgs(null as IEnumerable<AttackParamLocation>);
```
ditto

derevnjuk commented on src/SecTester.Scan/Target/Har/Response.cs:
```diff
@@ -0,0 +1,8 @@
+
+public record Response(Content Content) : Message
+{
+  public int Status { get; init; } = 200;
+  public string StatusText { get; init; } = "OK";
+  public string RedirectUrl { get; init; } = "";
```
No such default values according to spec or even sense

derevnjuk commented on src/SecTester.Scan/Target/Har/Entry.cs:
```diff
@@ -0,0 +1,12 @@
+
+public record Entry(DateTime StartedDateTime)
+{
+  public int Time { get; init; }
```
```suggestion
  public int Time { get; init; } = 0
```

derevnjuk commented on src/SecTester.Scan/Target/Har/Entry.cs:
```diff
@@ -0,0 +1,12 @@
+public record Entry(DateTime StartedDateTime)
+{
+  public int Time { get; init; }
+  public Timings? Timings { get; init; }
+  public Cache? Cache { get; init; }
+  public Request? Request { get; init; }
+  public Response? Response { get; init; }
```
All of them are mandatory fields according to spec

derevnjuk commented on src/SecTester.Scan/Target/Har/Content.cs:
```diff
@@ -0,0 +1,3 @@
+namespace SecTester.Scan.Target.Har;
+
+public record Content(int Size = -1, string MimeType = "text/plain");
```
ditto

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,166 @@
+  {
+    Target = targetOptions;
+    Tests = tests;
+    Name = CreateDefaultName(_target!);
```
```suggestion
    Name = CreateDefaultName(_target);
```

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,166 @@
+  {
+    Target = scanSettingsOptions.Target;
+    Tests = scanSettingsOptions.Tests;
+    Name = string.IsNullOrWhiteSpace(scanSettingsOptions.Name) ? CreateDefaultName(_target!) : scanSettingsOptions.Name;
```
```suggestion
    Name = string.IsNullOrWhiteSpace(scanSettingsOptions.Name) ? CreateDefaultName(_target) : scanSettingsOptions.Name;
```

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,166 @@
+    get => _poolSize;
+    init
+    {
+      if (value is null or < 1 or > 50)
```
Why did you remove a dedicated fields/properties for magic numbers?

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,166 @@
+    get => _targetTimeout;
+    init
+    {
+      if (value is null || (value < TimeSpan.FromSeconds(1) || value > TimeSpan.FromSeconds(120)))
```
ditto

derevnjuk commented on src/SecTester.Scan/ScanSettings.cs:
```diff
@@ -0,0 +1,166 @@
+    get => _slowEpTimeout;
+    init
+    {
+      if (value is null || value < TimeSpan.FromSeconds(100))
```
ditto

derevnjuk commented on src/SecTester.Scan/DefaultScanFactory.cs:
```diff
@@ -0,0 +1,111 @@
+    {
+      Time = 0,
+      Request = await target.ToHarRequest().ConfigureAwait(false),
+      Timings = new Timings { Send = 0, Receive = 0, Wait = 0 },
```
```suggestion
      Timings = new Timings(),
```

derevnjuk commented on src/SecTester.Scan/DefaultScanFactory.cs:
```diff
@@ -0,0 +1,111 @@
+        HttpVersion = "HTTP/1.1",
+        Status = 200,
+        StatusText = "OK",
+        HeadersSize = -1,
+        BodySize = -1,
+        RedirectUrl = "",
+        Cookies = new List<Cookie>(),
+        Headers = new List<Header>()
```
All of them have default values

```suggestion
        RedirectUrl = ""
```

derevnjuk commented on src/SecTester.Scan/DefaultScanFactory.cs:
```diff
@@ -0,0 +1,111 @@
+  {
+    return new Entry(_systemTimeProvider.Now)
+    {
+      Time = 0,
```
ditto

ostridm committed:
feat(scan): provide a default scan factory

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040790591
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040792843
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040793854
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040797477
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040801133
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040804097
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040806417
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040815093
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040815218
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040816056
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040819730
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040822333
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040830418
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040830723
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040831079
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040838219
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040855648
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040855892
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040860833
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040829888
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040819158
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040817408
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040828988
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040831709
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040833247
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040833813
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040833594
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040857798
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040857890
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040857928
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040822593
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040823438
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040826770
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040863325
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040863458
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040836299
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040839898
https://github.com/NeuraLegion/sectester-net/pull/76#discussion_r1040839982

closes #68

derevnjuk committed:
docs(scan): improve code snippets and examples

closes #68

derevnjuk committed:
refactor(scan): rename `Har` folder to `HarSpec`

closes #68

derevnjuk committed:
style: remove redundant trailing newlines

closes #68
